api/routes/sample.py

```diff
@@ -60,6 +75,37 @@ async def create_new_sample(
         return sample_id_format(internal_id)
 
 
+@router.put(
+    '/{project}/batch', response_model=Dict[str, Any], operation_id='batchUpsertSamples'
+)
+async def batch_upsert_samples(
+    samples: SampleBatchUpsert,
+    connection: Connection = get_project_write_connection,
+) -> Dict[str, Any]:
+    """Upserts a list of samples with sequences, and returns the list of internal sample IDs"""
+
+    # Table interfaces
+    st = SampleLayer(connection)
+    seqt = SampleSequenceLayer(connection)
+
+    async with connection.connection.transaction():
+        # Create or update samples
+        iids = [await st.upsert_sample(s) for s in samples.samples]
+        sids = [sample_id_format(x) for x in iids]
+
+        # Upsert all sequences with paired sids
+        sequences = zip(sids, [x.sequences for x in samples.samples])
+        seqs = [await seqt.upsert_sequences(sid, seqs) for sid, seqs in sequences]
+
+        # Format and return response
+        seqs_w_ids = dict(zip(iids, seqs))
+        results = [
+            {'sample_id': sample_id_format(k), 'sequences': v}
+            for k, v in seqs_w_ids.items()
+        ]
+        return dict(zip(iids, results))
```

Comment:
`SampleLayer`? Happy to move it, just not sure what you mean.

Reply:
Creating a function called `upsert_batch` in SampleLayer that takes the `SampleBatchUpsert` and creates the transaction and calls the upsert_samples and upsert_sequences. So if in future someone produces a SampleBatchUpsert model, they can call that method directly, rather than this logic in the route.